primitives/core/src/crypto.rs

```diff
@@ -354,278 +350,27 @@ fn ss58hash(data: &[u8]) -> blake2_rfc::blake2b::Blake2bResult {
-				match self {
-					$(
-						Ss58AddressFormat::$identifier => write!(f, "{}", $name),
-					)*
-					Ss58AddressFormat::Custom(x) => write!(f, "{}", x),
-				}
-
-			}
-		}
-
-		static ALL_SS58_ADDRESS_FORMATS: [Ss58AddressFormat; 0 $(+ { let _ = $number; 1})*] = [
-			$(Ss58AddressFormat::$identifier),*,
-		];
-
-		impl Ss58AddressFormat {
-			/// names of all address formats
-			pub fn all_names() -> &'static [&'static str] {
-				&[
-					$($name),*,
-				]
-			}
-			/// All known address formats.
-			pub fn all() -> &'static [Ss58AddressFormat] {
-				&ALL_SS58_ADDRESS_FORMATS
-			}
-
-			/// Whether the address is custom.
-			pub fn is_custom(&self) -> bool {
-				matches!(self, Self::Custom(_))
-			}
-		}
-
-		impl TryFrom<u8> for Ss58AddressFormat {
-			type Error = ();
-
-			fn try_from(x: u8) -> Result<Ss58AddressFormat, ()> {
-				Ss58AddressFormat::try_from(x as u16)
-			}
-		}
-
-		impl From<Ss58AddressFormat> for u16 {
-			fn from(x: Ss58AddressFormat) -> u16 {
-				match x {
-					$(Ss58AddressFormat::$identifier => $number),*,
-					Ss58AddressFormat::Custom(n) => n,
-				}
-			}
-		}
-
-		impl TryFrom<u16> for Ss58AddressFormat {
-			type Error = ();
-
-			fn try_from(x: u16) -> Result<Ss58AddressFormat, ()> {
-				match x {
-					$($number => Ok(Ss58AddressFormat::$identifier)),*,
-					_ => Ok(Ss58AddressFormat::Custom(x)),
-				}
-			}
-		}
-
-		/// Error encountered while parsing `Ss58AddressFormat` from &'_ str
-		/// unit struct for now.
-		#[derive(Copy, Clone, PartialEq, Eq, crate::RuntimeDebug)]
-		pub struct ParseError;
-
-		impl<'a> TryFrom<&'a str> for Ss58AddressFormat {
-			type Error = ParseError;
-
-			fn try_from(x: &'a str) -> Result<Ss58AddressFormat, Self::Error> {
-				match x {
-					$($name => Ok(Ss58AddressFormat::$identifier)),*,
-					a => a.parse::<u16>().map(Ss58AddressFormat::Custom).map_err(|_| ParseError),
-				}
-			}
-		}
-
-		#[cfg(feature = "std")]
-		impl std::str::FromStr for Ss58AddressFormat {
-			type Err = ParseError;
-
-			fn from_str(data: &str) -> Result<Self, Self::Err> {
-				Self::try_from(data)
-			}
-		}
-
-		#[cfg(feature = "std")]
-		impl std::fmt::Display for ParseError {
-			fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-				write!(f, "failed to parse network value as u8")
-			}
-		}
+ss58_registry::ss58_registry!();
 
-		#[cfg(feature = "std")]
-		impl Default for Ss58AddressFormat {
-			fn default() -> Self {
-				*DEFAULT_VERSION.lock()
-			}
-		}
+/// Default prefix number
+#[cfg(feature = "std")]
+static DEFAULT_VERSION: core::sync::atomic::AtomicU16 = core::sync::atomic::AtomicU16::new(42 /*substrate*/);
 
-		#[cfg(feature = "std")]
-		impl From<Ss58AddressFormat> for String {
-			fn from(x: Ss58AddressFormat) -> String {
-				x.to_string()
-			}
-		}
-	)
+#[cfg(feature = "std")]
+impl Default for Ss58AddressFormat {
+	fn default() -> Self {
+		DEFAULT_VERSION.load(core::sync::atomic::Ordering::Relaxed).into()
+	}
 }
```

Comment:
Where do we need the default?

I think it would still be nice to get rid off this. We could have a function `ss58_address_format()` here in Substrate that does this.

Reply:
There's ~9 places. I've re-pointed them to equivalents. Have made the enum be in its own crate now. Keen on your thoughts - is this more what you were thinking of?